Option to filter posts and comments on user profile page

pazoozoo42 opened the pull request:
This is related to search so I started from the `search` branch

https://user-images.githubusercontent.com/57953386/208183015-52ba7436-9e0d-46e5-ba03-80a2508d2895.mp4

Aivean commented on backend/src/db/repositories/CommentRepository.ts:
```diff
@@ -43,23 +43,36 @@ export default class CommentRepository {
+              limit_from: limitFrom,
+              limit_count: perPage
+          };
+        if (filter) {
```
The queries are mostly identical, should we merge them?

pazoozoo42 replied:
done

Aivean commented on backend/src/db/repositories/CommentRepository.ts:
```diff
@@ -43,23 +43,36 @@ export default class CommentRepository {
+                    from comments c
+                      left join comment_votes v on (v.comment_id = c.comment_id and v.voter_id = :for_user_id)
+                    where
+                      c.author_id = :user_id and c.deleted = 0 and c.source like :filter
```
Let's maybe use full text index here?

pazoozoo42 replied:
I am unsure whether we should use it here and I am debating with myself, but I tend to lean towards not using it. This would duplicate advanced search functionality once it's there

Aivean replied:
Ok, if that's just a temporary solution to be replaced with elasticsearch, I'm ok with it.

Aivean commented on backend/src/db/repositories/PostRepository.ts:
```diff
@@ -83,28 +83,46 @@ export default class PostRepository {
+            limit_count: perPage
+        };
+        if (filter) {
+            query = `select p.*, v.vote, b.read_comments, b.bookmark, b.last_read_comment_id, b.watch
```
Same as above, let's try to avoid the query duplication.

pazoozoo42 replied:
I am unsure whether we should use it here and I am debating with myself, but I tend to lean towards not using it. This would duplicate advanced search functionality once it's there

Aivean reviewed (commented):
Looks good overall! And it's a very useful feature, thank you for working on it.

UX related thoughts: perhaps the search icon will look better instead of the text? Or, alternatively, we can get rid of the button altogether, replacing it with a debounced request on input change.

pazoozoo42 commented:
> Or, alternatively, we can get rid of the button altogether, replacing it with a debounced request on input change.

implemented "search as you type", removed the button

Aivean commented on backend/src/db/repositories/CommentRepository.ts:
```diff
@@ -43,23 +43,27 @@ export default class CommentRepository {
+              limit_count: perPage
+          };
+        if (filter) {
+            params['filter'] = '%' + filter + '%';
```
We should test this for potential abuse. E.g. super long queries with `%` in them.

Aivean replied:
Fixed.

Aivean commented on backend/src/db/repositories/CommentRepository.ts:
```diff
@@ -68,10 +72,13 @@ export default class CommentRepository {
+    async getUserCommentsTotal(userId: number, filter: string): Promise<number> {
+        const query = `select count(*) cnt from comments where author_id = ? and deleted = 0 ${filter ? '  and source like ? ' : ''}`;
+        const params: Array<number | string> = [userId];
+        if (filter) {
```
Regarding the conditional params. I just tested, and it seems that you can have unused params with `undefined` value:

```
{
              username: username,
              'test': undefined
          }
```
Should reduce the boilerplate.

Aivean replied:
Fixed.

Aivean reviewed (approved):
I addressed my review comments. Feel free to merge it.

One note though:  We need to address the current filter inefficiency before releasing this, either by switching to the full text index, or by using elasticsearch. Currently it's a performance concern.
